File: source/downloads/code/LeetCode/Java/Valid-Sudoku.java
public class Solution {
    public boolean isValidSudoku(char[][] board) {
        int row = board.length;
        int col = board[0].length;

        for(int i=0;i<row;i++){
            for (int j=0;j<col;j++){
                if (board[i][j]=='.') continue;
            
                // 行
                int k=0;
                while (k<9){
                    if(j!=k && board[i][k] == board[i][j])
                        return false;
                    k+=1;    
                }
            
                // 列
                k = 0;
                while(k<9){
                    if (i!=k && board[k][j] == board[i][j])
                        return false;
                    k+=1;
                }

                // ３×３
                for (int m=0;m<3;m++){
                    for (int n=0;n<3;n++){
                        if (i/3*3+m !=i && j/3*3+n != j && board[i][j] == board[i/3*3+m][j/3*3+n])
                            return false;
                    }
                }
            }
        }
        return true;       
    }
}

File: downloads/code/LeetCode/Python/Valid-Sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        
        def check(x,y,board):
            # 检查横行是否和board[x][y]重复
            for i in range(9):
                if y != i and board[x][i] == board[x][y]:
                    return False

            # 检查纵行是否和board[x][y]重复
            for i in range(9):
                if x != i and board[i][y] == board[x][y]:
                    return False

            # 检查3*3的小矩阵是否和board[x][y]重复
            for i in range(3):
                for j in range(3):
                    # x/3 找到是第几个3*3的矩阵
                    if x/3*3+i!=x and y/3*3+j!=y and board[x/3*3+i][y/3*3+j] == board[x][y]:
                        return False
            return True

    
        for i in range(9):
            for j in range(9):
                if board[i][j] == ".":
                    continue
                if check(i,j,board) == False:
                    return False
        return True


